compiler/src/symboltable.rs

```diff
@@ -193,71 +200,137 @@ impl<'a> SymbolTableAnalyzer<'a> {
+                        if !symbols.contains_key(&symbol.name) || scope_depth < 2 {
+                            return Err(SymbolTableError {
+                                error: format!("no binding for nonlocal '{}' found", symbol.name),
+                                location: Default::default(),
+                            });
+                        }
+                    } else {
                         return Err(SymbolTableError {
-                            error: format!("no binding for nonlocal '{}' found", symbol.name),
+                            error: format!(
+                                "nonlocal {} defined at place without an enclosing scope",
+                                symbol.name
+                            ),
                             location: Default::default(),
                         });
                     }
-                } else {
-                    return Err(SymbolTableError {
-                        error: format!(
-                            "nonlocal {} defined at place without an enclosing scope",
-                            symbol.name
-                        ),
-                        location: Default::default(),
-                    });
+                }
+                SymbolScope::Global => {
+                    // TODO: add more checks for globals?
+                }
+                SymbolScope::Local => {
+                    // all is well
+                }
+                SymbolScope::Unknown => {
+                    // Try hard to figure out what the scope of this symbol is.
+                    self.analyze_unknown_symbol(symbol);
                 }
             }
-            SymbolScope::Global => {
-                // TODO: add more checks for globals?
+        }
+        Ok(())
+    }
+
+    fn analyze_unknown_symbol(&self, symbol: &mut Symbol) {
+        if symbol.is_assigned || symbol.is_parameter {
+            symbol.scope = SymbolScope::Local;
+        } else {
+            // Interesting stuff about the __class__ variable:
+            // https://docs.python.org/3/reference/datamodel.html?highlight=__class__#creating-the-class-object
+            let found_in_outer_scope = symbol.name == "__class__"
+                || self.tables.iter().skip(1).any(|(symbols, typ)| {
+                    *typ != SymbolTableType::Class && symbols.contains_key(&symbol.name)
+                });
+
+            if found_in_outer_scope {
+                // Symbol is in some outer scope.
+                symbol.is_free = true;
+            } else if self.tables.is_empty() {
+                // Don't make assumptions when we don't know.
+                symbol.scope = SymbolScope::Unknown;
+            } else {
+                // If there are scopes above we can assume global.
+                symbol.scope = SymbolScope::Global;
             }
-            SymbolScope::Local => {
-                // all is well
+        }
+    }
+
+    fn analyze_symbol_comprehension(
```

Comment:
Could you explain the reasoning behind this function? I suspect it might be tricky to follow a different code path for symbols in named expressions, can the existing function `analyze_symbol` be modified to incorporate named expressions?

Could you maybe give an example python snippet why this function is required?

Reply:
When you use a name the first time in an named expression in a comprehension you need to propagate the name to the next level outside of the outermost comprehension. Due to the generic multilevel scoping issue in rustpython it is currently not that helpfull (RustPytohn supports only the directly containing level, what would be be sufficient with that limitation - I added a [testcase for reference](https://github.com/TheAnyKey/RustPython/blob/e1d382d8df6423f2da9878d6b066e9fb579d9b82/Lib/test/test_named_expression.py#L399)). In that way the implementation is complete and will work immediately after solving this.
This is inspired by the extend scope function https://github.com/python/cpython/blob/7b78e7f9fd77bb3280ee39fb74b86772a7d46a70/Python/symtable.c#L1435
thought some aspects in the error handling are still missing. As this would require more context information, e.g., we distinguish iterator variables from others, what´s currently not implemented.

Reply:
Okay, thank you for the explanation! Could we use the non-local in some way to fix this?

Reply:
Hi Windel,
with the non-local stuff we can only fix the part bring in the variable from the outer scope for reading. Writing wont work, therefore also the scope.rs needs to be fixed. Currently when storing to a non-local it takes the directly containing scope and writes the variable to this. This is too simple and does not work in general. Instead you also need to search for the next scope that contains the name,  as when reading from a non-local. Doing this in the straight forward way is stronger than whats now implemented and should mostly be sufficient for the pre-3.8 stuff. For the named expressions, we need to take more context into account to implement the scope extension for so not used names in comprehensions

Reply:
see #1951